### PR TITLE
Allow `Closed` pools to be opened

### PR DESCRIPTION
**Pull request by maltekliemann**
This adds the pool status `Initialized`, with which all Cpmm pool start, and adds the functionality to open an initialized pool.

The purpose of this is to allow `prediction-markets` to open and close pools depending on market status. The motivation for allowing this is to let a market which is created _before_ it opens to close the pool until the market opens to prevent trading. This will be implemented later. Note that currently, we still call `open_pool` when creating a cpmm pool by default.

### CODE REVIEW

File: zrml/swaps/src/tests.rs [hunk 1]
```diff
@@ -2475,6 +2542,9 @@ fn create_initial_pool(
         },
         if scoring_rule == ScoringRule::CPMM { Some(vec!(_2, _2, _2, _2)) } else { None },
     ));
+    if scoring_rule == ScoringRule::CPMM {
+        assert_ok!(Swaps::open_pool(pool_id));
```
[the remark this answers is not in the document]
Comment (Chralt98) [reply zrml/swaps/src/tests.rs:2546]: I am missing a test which checks the state `Initialized` before the `create_pool` call on CPMM.

File: zrml/swaps/src/tests.rs [hunk 2]
```diff
@@ -281,21 +294,21 @@ fn create_pool_generates_a_new_pool_with_correct_parameters_for_cpmm() {
...
-        assert_eq!(pool.total_weight.unwrap(), _8);
+        assert_eq!(pool.total_weight.unwrap(), _10);
 
-        assert_eq!(*pool.weights.as_ref().unwrap().get(&ASSET_A).unwrap(), _2);
-        assert_eq!(*pool.weights.as_ref().unwrap().get(&ASSET_B).unwrap(), _2);
+        assert_eq!(*pool.weights.as_ref().unwrap().get(&ASSET_A).unwrap(), _4);
+        assert_eq!(*pool.weights.as_ref().unwrap().get(&ASSET_B).unwrap(), _3);
         assert_eq!(*pool.weights.as_ref().unwrap().get(&ASSET_C).unwrap(), _2);
-        assert_eq!(*pool.weights.as_ref().unwrap().get(&ASSET_D).unwrap(), _2);
+        assert_eq!(*pool.weights.as_ref().unwrap().get(&ASSET_D).unwrap(), _1);
```
Comment (Chralt98) [inline zrml/swaps/src/tests.rs:304]: useful to have not a standard values when actually checking the pool parameters

**Review by Chralt98 — COMMENTED** [review]
Thanks for adding the `Initialized` state, so that trading is only allowed after the prediction market pallet allows it. I mentioned some small things.

File: zrml/swaps/src/tests.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (maltekliemann) [reply zrml/swaps/src/tests.rs:2546]: `create_pool` has the same accessibility as `open_pool`. So any user that creates a pool manually (a pallet, for example) is also capable of opening it. But this entire control flow is hidden from the users.

**Review by Chralt98 — APPROVED** [review]
Looks like a charm.

**Review by vivekvpandya — APPROVED** [review]

**Review by vivekvpandya — APPROVED** [review]

**Review by Chralt98 — APPROVED** [review]
Ok, as you pointed out in Discord: it needs the state of either `Initialized` or `Active` (so preventing `Closed` and `Clean`) to join the pool. I couldn't find anything to update. What happens after cleaning up the pool. What happens when I call `pool_exit` then?

**Comment by maltekliemann** [comment]
`exit_pool` is always ok. The pool (usually) goes into `Clean` state when the market owning the pools resolves to a certain outcome. Even then (especially then!) the LPs should be allowed to withdraw their funds.

**Comment by maltekliemann** [comment]
Thanks for the review, guys! :thumbsup:

File: zrml/swaps/src/lib.rs [hunk 3]
```diff
@@ -370,6 +370,10 @@ mod pallet {
         ) -> DispatchResultWithPostInfo {
             let who = ensure_signed(origin)?;
             let pool = Self::pool_by_id(pool_id)?;
+            ensure!(
+                matches!(pool.pool_status, PoolStatus::Initialized | PoolStatus::Active),
+                Error::<T>::InvalidPoolStatus,
+            );
```
[the remark this answers is not in the document]
Comment (maltekliemann) [reply zrml/swaps/src/lib.rs:376]: Yeah, after that, joining doesn't really make any sense.

**Pull request merged**